Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialPropertyConnectionSerializer.cpp

```diff
@@ -0,0 +1,126 @@
+                static constexpr const char id[] = "id"; // For backward compatibility
+                static constexpr const char shaderIndex[] = "shaderIndex";
+            }
+
+            static const AZStd::string_view AcceptedFields[] =
+            {
+                Field::type,
+                Field::name,
+                Field::id,
+                Field::shaderIndex
+            };
+        }
+
+        AZ_CLASS_ALLOCATOR_IMPL(JsonMaterialPropertyConnectionSerializer, SystemAllocator, 0);
+
+        JsonSerializationResult::Result JsonMaterialPropertyConnectionSerializer::Load(void* outputValue, const Uuid& outputValueTypeId,
+            const rapidjson::Value& inputValue, JsonDeserializerContext& context)
+        {
+            namespace JSR = JsonSerializationResult;
+            using namespace JsonMaterialPropertyConnectionSerializerInternal;
+
+            AZ_Assert(azrtti_typeid<MaterialTypeSourceData::PropertyConnection>() == outputValueTypeId,
+                "Unable to deserialize material property connection to json because the provided type is %s",
+                outputValueTypeId.ToString<AZStd::string>().c_str());
+            AZ_UNUSED(outputValueTypeId);
+
+            MaterialTypeSourceData::PropertyConnection* propertyConnection = reinterpret_cast<MaterialTypeSourceData::PropertyConnection*>(outputValue);
+            AZ_Assert(propertyConnection, "Output value for JsonMaterialPropertyConnectionSerializer can't be null.");
+
+            JSR::ResultCode result(JSR::Tasks::ReadField);
+
+            if (!inputValue.IsObject())
+            {
+                return context.Report(JsonSerializationResult::Tasks::ReadField, JsonSerializationResult::Outcomes::Unsupported, "Property connection must be a JSON object.");
+            }
+            
+            MaterialUtils::CheckForUnrecognizedJsonFields(AcceptedFields, AZ_ARRAY_SIZE(AcceptedFields), inputValue, context, result);
+
+            result.Combine(ContinueLoadingFromJsonObjectField(&propertyConnection->m_type, azrtti_typeid<MaterialPropertyOutputType>(), inputValue, Field::type, context));
+            
+            JsonSerializationResult::ResultCode nameResult = ContinueLoadingFromJsonObjectField(&propertyConnection->m_fieldName, azrtti_typeid<AZStd::string>(), inputValue, Field::name, context);
+            if (nameResult.GetOutcome() == JsonSerializationResult::Outcomes::DefaultsUsed)
+            {
+                // This "id" key is for backward compatibility.
+                result.Combine(ContinueLoadingFromJsonObjectField(&propertyConnection->m_fieldName, azrtti_typeid<AZStd::string>(), inputValue, Field::id, context));
+            }
+            else
+            {
+                result.Combine(nameResult);
+            }
+
+            result.Combine(ContinueLoadingFromJsonObjectField(&propertyConnection->m_shaderIndex, azrtti_typeid<int32_t>(), inputValue, Field::shaderIndex, context));
```

Comment:
I understand the point you made offline that this is still quite verbose for something simple. Even if there's a tool in the future to deal with the exact versioning problem, there will still be cases where only one or so fields in an object needs special handling with the rest needing to be simple. Seems like some form of batch version of `ContinueLoadingFromJsonObjectField` would be useful, so it's possible to declare something like:
```
auto batch[] = 
{
    BatchObject("a", instance->m_a),
    BatchObject("b", instance->m_b),
    BatchObject("c", instance->m_c)
}
result.Combine(ContinueLoadingFromJsonObjectField(batch, inputValue, context));
```
Same for storing off course. That could take care of at least the objects that just need to be loaded as normal.
(Come to think of it, some light type deduction through something like `BatchObject` or directly in an overload of `ContinueLoadingFromJsonObjectField` would help avoid pointers being passed in without the correct continuation flag set.)